app/models/term.rb

```diff
@@ -0,0 +1,105 @@
+class Term < ApplicationRecord
+  TEMPORARY_URI_BASE = 'temp:'.freeze
+  LOCAL     = 'local'.freeze
+  TEMPORARY = 'temporary'.freeze
+  EXTERNAL  = 'external'.freeze
+
+  TERM_TYPES = [LOCAL, EXTERNAL, TEMPORARY].freeze
+
+  VALID_URI_REGEX = /\A#{URI.regexp}\z/
+
+  belongs_to :vocabulary
+
+  before_validation :add_uuid, :add_uri, :add_uri_hash, on: :create
+
+  after_commit :update_solr # Is triggered after successful save/update/destroy.
+
+  validates :vocabulary, :pref_label, :uri, :uri_hash, :uuid, :term_type, presence: true
+  validates :term_type, inclusion: { in: TERM_TYPES }
+  validate  :uuid_uri_and_term_type_unchanged
+
+  store :custom_fields, coder: JSON
+
+  serialize :alt_label, Array
+
+  def to_solr
+    {
+      'uuid'          => uuid,
+      'uri'           => uri,
+      'pref_label'    => pref_label,
+      'alt_label'     => alt_label,
+      'term_type'     => term_type,
+      'vocabulary'    => vocabulary.string_key,
+      'authority'     => authority,
+      'custom_fields' => custom_fields.to_json
+    }.tap do |doc|
+      vocabulary.custom_fields.each do |k, v|
+        doc["#{k}#{URIService.solr_suffix(v[:data_type])}"] = self.custom_fields[k]
+      end
+    end
+  end
+
+  def add_custom_field(field, value)
```

Comment:
It might be more accurate to call this `set_custom_field` because it replaces an existing value rather than adding.

Reply:
Sure.